Polish the button

jasmussen opened the pull request:
This fixes #3283.

This is a redesign of the button block to be more neutral looking. In addition to this, it makes the URL aspect inline, so it behaves just like how a caption does on an image.

GIF:

![button](https://user-images.githubusercontent.com/1204802/32661648-032c970c-c628-11e7-8217-8b3721299ced.gif)

Commit:
Polish the button

This fixes #3283.

This is a redesign of the button block to be more neutral looking. In addition to this, it makes the URL aspect inline, so it behaves just like how a caption does on an image.

youknowriad commented on blocks/library/button/index.js:
```diff
@@ -130,6 +119,18 @@ registerBlockType( 'core/button', {
 			</span>,
+			focus && (
+				<form
+					className="blocks-format-toolbar__inline-link-modal"
```
I think we do not want to share the style of this class with other blocks, it may make sense to rename the classname to match our guidelines. Something like `blocks-button__inline-link`. Ans style it as part of the `editor.scss` of the button block

youknowriad commented:
Nice job, I like this new design way better. 👍 

I noticed some small design bugs

<img width="348" alt="screen shot 2017-11-10 at 16 13 33" src="https://user-images.githubusercontent.com/272444/32664775-8f3464aa-c632-11e7-89fb-125fa90e090b.png">

The submit button is crossing the border when the button is left/right aligned

<img width="708" alt="screen shot 2017-11-10 at 16 13 17" src="https://user-images.githubusercontent.com/272444/32664782-921d902e-c632-11e7-87f7-958d62b8af4b.png">

Should we align the link form on the left when the button is not centered?

jasmussen commented:
Thanks for the review, good stuff all around. I'll fix it next week. Any idea why Travis is complaining btw? Also, have a nice weekend!

youknowriad commented on blocks/library/button/index.js:
```diff
@@ -130,6 +119,18 @@ registerBlockType( 'core/button', {
+					/>
+					<IconButton icon="editor-break" label={ __( 'Apply' ) } type="submit" />
+				</form>
+			)
```
@jasmussen Yep, travis is failing because of a missing trailing comma after this parathensis. You should definitely install an `eslint` plugin to your favorite IDEE ;)

jasmussen replied:
💓💓👍🌌

youknowriad replied:
Have a nice weekend Joen :)

karmatosed commented:
I will hold off my design review until you have done those fixes @jasmussen - just let me know when want one.

Commit:
Address feedback

- Move to button-specific class
- Align url to match button alignment
- Fix input field overlap

Commit:
Fix linting issue

jasmussen commented:
Thanks again, ready for final review now. I've addressed the feedback and fixed the linting issue. Can you recommend a linter for Atom? The last one I used sort of broke.

Screenshots:

![screen shot 2017-11-13 at 09 48 42](https://user-images.githubusercontent.com/1204802/32716930-26a3c0de-c858-11e7-98a4-0441ac047762.png)

![screen shot 2017-11-13 at 09 48 54](https://user-images.githubusercontent.com/1204802/32716933-2801d768-c858-11e7-99ef-b1c776537f84.png)

youknowriad commented:
Based on quick search, this looks like the widely used eslint atom plugin https://atom.io/packages/linter-eslint (I don't use atom myself though)

youknowriad reviewed (approved):
LGTM 👍

hedgefield commented:
Good stuff! One question and one remark:

- What does the `Stand on a line` option in the block settings mean?
- The Apply button should have some sort of "this has been successfully applied" feedback. When I entered a URL and pressed Enter, nothing seemed to change, so I clicked the button, and still nothing visibly changed. It would be good to hide (or at least dim) the Apply button once you've successfully entered something, and only show it again when you change something in the URL.

jasmussen commented:
> What does the Stand on a line option in the block settings mean?

Perhaps it would be nice to be able to show little ? help icons, with elaborating tooltips. In this case, "Stand on a line" is basically the same as applying `overflow: hidden;` to the block container. It prevents successive floated blocks from stacking up to the right of a left floated button. It's a "float clearer", in other words. Perhaps we should open a separate ticket for little help icons with tooltips?

> The Apply button should have some sort of "this has been successfully applied" feedback. 

I don't disagree. Can you think of other areas where an "applied" affordance would be helpful? The pattern emulated with this link was the caption on images. If we can think of a few places, it might be good to ticket separately and then reference all the places where such an affordance would help.

hedgefield commented:
Ahhh, that's something I would call "full width" :) so coming up with a good general term for that would be useful. It's an issue with images too, though not so much the language there, but the concept, the difference between full width and center aligned: https://github.com/WordPress/gutenberg/issues/3325. If we can come up with a good term we can use the 'stand on a line' toggle to solve the problem there too. 👍 

Making a separate issue for the Apply button is a good idea. Off the top of my head, it is used in links, captions, buttons, permalink.. cover image has a link field too, maybe the shortcode block.

karmatosed commented:
>  Perhaps we should open a separate ticket for little help icons with tooltips?

Just a little note, I am cautious of us adding help icons. I am totally into us being clearer, but if we have to add a help icon maybe we aren't being in first place? That said, help has a place, it should always though mean we look first at if we can not add it.

I do think 'Stand on a line' could do with some work. Maybe literally say what it does?

"Ensure clears to new line" or something a little less clumsy.

As for an applied notice, I agree, it's probably a useful feedback mechanism. I started a ticket to consider the possible areas this could be. #3461
